crates/sui-framework/packages/sui-framework/sources/k_of_n_upgrade_policy.move

```diff
@@ -0,0 +1,522 @@
+    }
+
+    /// Record the transfer for a `Ballot`.
+    struct BallotTransfered has copy, drop {
+        /// the ID of the proposal (`ProposedUpgrade` instance)
+        proposal: ID,
+        /// digest of the proposal
+        digest: vector<u8>,
+        /// ballot used for this vote
+        ballot: ID,
+        /// signer of the transfer operation
+        signer: address,
+        /// original owner of the ballot, that is the address the
+        /// ballot was first sent to
+        original_owner: address,
+        /// the receiver of the ballot
+        ballot_receiver: address,
+    }
+
+    /// Record a deleted `Ballot`.
+    struct BallotDeleted has copy, drop {
+        /// the ID of the proposal (`ProposedUpgrade` instance)
+        proposal: ID,
+        /// digest of the proposal
+        digest: vector<u8>,
+        /// ballot used for this vote
+        ballot: ID,
+        /// signer of the deleted operation
+        signer: address,
+        /// original owner of the ballot, that is the address the
+        /// ballot was first sent to
+        original_owner: address,
+    }
+
+    /// Allowed voters must in the [2, 100] range.
+    const EAllowedVotersError: u64 = 0;
+    /// Required votes must be less than allowed voters.
+    const ERequiredVotesError: u64 = 1;
+    /// The signer of the proposed upgrade is not one of the allowed voters.
+    const EInvalidProposer: u64 = 2;
+    /// The `Ballot` used to vote is not for the correct proposal (`ProposedUpgrade`)
+    const EInvalidBallot: u64 = 3;
+    /// Not enough votes to perform the upgrade.
+    const ENotEnoughVotes: u64 = 4;
+    /// An upgrade was issued already, and the operation requested failed.
+    const EAlreadyIssued: u64 = 5;
+    /// The operation required the signer to be the same as the upgrade proposer.
+    const ESignerMismatch: u64 = 6;
+    /// Proposal (`KofNUpgradeCap`) and upgrade (`ProposedUpgrade`) do not match.
+    const EInvalidProposalForUpgrade: u64 = 7;
+
+    /// Create a `KofNUpgradeCap` given an `UpgradeCap`.
+    /// The returned instance is the only and exclusive controller of upgrades. 
+    /// If the transaction sender is not in the set of allowed voters it will be
+    /// added. The sender is the publisher of the package which is always allowed
+    /// to vote.
+    /// The `k` (`required_votes`) out of `n` (length of `voters`) is set up
+    /// at construction time and it is immutable.
+    public fun new(
+        cap: UpgradeCap,
+        required_votes: u64,
+        voters: VecSet<address>,
+        ctx: &mut TxContext,
+    ): KofNUpgradeCap {
+        assert!(vec_set::size(&voters) > 1, EAllowedVotersError);
+        // add the signer to the list of voters if that was not there
+        let signer = tx_context::sender(ctx);
+        if (!vec_set::contains(&voters, &signer)) {
+            vec_set::insert(&mut voters, signer);
+        };
+        assert!(vec_set::size(&voters) <= 100, EAllowedVotersError);
```

Comment:
Why do we care how many voters there are?

Reply:
Is this something about object size or something?

Reply:
Fair!

Reply:
we do not necessarily care but there are operations that get more expensive if we allow a big enough number.
Also if we stick with this model there would be a ton of object creation that could get pricey.
And it's really unclear what we are trying to accomplish with a very general solution. I would go even down from 100. 
I think it's clear this is not intended to be the one and only k out of n policy we publish.

Reply:
resolving because it is getting a bit too messy otherwise.
Basically the limit is sane considering we "distribute" `VotingCap` in the creation of the policy and all solutions we are considering require creation of objects.
It seems just a waste and a possible problem (hitting limits) to have the number being unlimited for now

Reply:
I think we should also allow `voters` to be `1` as well (to support a 1-of-1 policy, which acts as a form of delegation -- i.e. one address holds the cap, and can request an upgrade, while another has to give permission).

Reply:
done